README.rst

```diff
@@ -1962,6 +1968,341 @@ However, if specifying :code:`Table` objects manually, please make sure to inclu
+
+One of our options is to materialize such an edge in the underlying database itself.
+However, this causes denormalization of the database -- there are now two places where
+an animal's grandchildren are written down -- requiring additional storage space,
+and introducing potential for user confusion and data inconsistency between the two representations.
+
+Another option is to introduce a non-materialized view within the database that *makes it appear*
+that such an edge exists, and query this view via the GraphQL compiler. While this avoids some
+of the drawbacks of the previous approach, not all databases support non-materialized views.
+Also, querying users are not always able to add views to the database, and may require additional
+permissions on the database system.
+
+Macro edges give us the opportunity to define a new :code:`out_Animal_GrandparentOf` edge without
+involving the underlying database systems at all. We simply state that such an edge
+is constructed by composing two :code:`out_Animal_ParentOf` edges together:
+
+.. code:: python
+
+    from graphql_compiler.macros import register_macro_edge
+
+    macro_edge_definition = '''{
+        Animal @macro_edge_definition(name: "out_Animal_GrandparentOf") {
+            out_Animal_ParentOf {
+                out_Animal_ParentOf @macro_edge_target {
+                    uuid
+                }
+            }
+        }
+    }'''
+    macro_edge_args = {}
+
+    register_macro_edge(your_macro_registry_object, macro_edge_definition, macro_edge_args)
+
+Let's dig into the GraphQL macro edge definition one step at a time:
+
+- We know that the new macro edge is being defined on the :code:`Animal` GraphQL type, since that
+  is the type where the definition begins.
+
+- The :code:`@macro_edge_definition` directive specifies the name of the new macro edge.
+
+- The newly-defined :code:`out_Animal_GrandparentOf` edge connects :code:`Animal` vertices
+  to the vertices reachable after exactly two traversals along :code:`out_Animal_ParentOf` edges;
+  this is what the :code:`@macro_edge_target` directive signifies.
+
+- As the :code:`out_Animal_ParentOf` field containing the :code:`@macro_edge_target` directive
+  is of type :code:`[Animal]` (we know this from our schema), the compiler will automatically infer
+  that the :code:`out_Animal_GrandparentOf` macro edge also points to vertices
+  of type :code:`Animal`.
+
+- The :code:`uuid` within the inner :code:`out_Animal_ParentOf` scope is a "pro-forma" field -- it
+  is there simply to satisfy the GraphQL parser, since per the GraphQL specification, each pair of
+  curly braces must reference at least one field. The named field has no meaning in this definition,
+  and the user may choose to use any field that exists within that pair of curly braces.
+  The preferred convention for pro-forma fields is to use the field of GraphQL type :code:`ID`
+  which is the primary key of the given type in the underlying database.
```

Comment:
The ID type does not need to be the primary key in the underlying database. In fact, the ID type has a python string type and in the cases where the primary key is an int, it can't be a primary key.

Reply:
Removed the distracting `ID` type reference. A well-formed GraphQL schema will have the primary key be typed `ID` for compliance with GraphQL principles, but the macro system can be more lenient.